### ISSUE TITLE
Bug: Kubernetes DNS configuration is broken after switching from v3.8.1 to gluetun:latest

### ISSUE THREAD

**Issue by npawelek** [qdm12/gluetun#326]
**Host OS**: Ubuntu 20.04

**Is this urgent?**: No, I'll remain on the older version for the time being.

**What VPN provider are you using**: PIA

**What are you using to run your container?**: Kubernetes

**What is the version of the program** (See the line at the top of your logs)

```
Running version latest built on 2020-12-30T15:54:51Z (commit fb8a615)
```

**What's the problem**: I attempt to pin my container DNS using `DNS_PLAINTEXT_ADDRESS` which is working fine for me in v3.8.1 (I don't use DOT through the container). When I switched to the new `qmcgaw/gluetun` container and updated env vars to match latest, I'm seeing a few issues:

1. Errors with unbound failing to access the defined DNS server. As this was working in the prior version, I suspect the issue is with the firewall preventing this connection out, even though I have the firewall vars configured (`FIREWALL_OUTBOUND_SUBNETS` defined with `10.96.0.0/12`). This is working fine in the previous version. Here's what I see in the logs with unbound:

```bash
2020-12-30T11:16:30.022-0600    INFO    dns configurator: downloading root hints from https://raw.githubusercontent.com/qdm12/files/master/named.root.updated
2020-12-30T11:16:30.360-0600    INFO    dns configurator: downloading root key from https://raw.githubusercontent.com/qdm12/files/master/root.key.updated
2020-12-30T11:16:30.431-0600    INFO    dns configurator: generating Unbound configuration
2020-12-30T11:16:30.431-0600    INFO    dns configurator: 0 hostnames blocked overall
2020-12-30T11:16:30.431-0600    INFO    dns configurator: 0 IP addresses blocked overall
2020-12-30T11:16:30.431-0600    INFO    dns configurator: starting unbound
2020-12-30T11:16:30.431-0600    INFO    dns configurator: using DNS address 127.0.0.1 internally
2020-12-30T11:16:30.431-0600    INFO    dns configurator: using DNS address 127.0.0.1 system wide
2020-12-30T11:16:30.488-0600    INFO    unbound: start of service (unbound 1.10.1).
2020-12-30T11:16:30.734-0600    WARN    dns configurator: could not resolve github.com (try 1 of 6): lookup github.com on 10.96.0.10:53: server misbehaving
2020-12-30T11:16:30.837-0600    WARN    dns configurator: could not resolve github.com (try 2 of 6): lookup github.com on 10.96.0.10:53: server misbehaving
2020-12-30T11:16:31.139-0600    WARN    dns configurator: could not resolve github.com (try 3 of 6): lookup github.com on 10.96.0.10:53: server misbehaving
2020-12-30T11:16:31.642-0600    WARN    dns configurator: could not resolve github.com (try 4 of 6): lookup github.com on 10.96.0.10:53: server misbehaving
2020-12-30T11:16:32.644-0600    WARN    dns configurator: could not resolve github.com (try 5 of 6): lookup github.com on 10.96.0.10:53: server misbehaving
2020-12-30T11:16:34.647-0600    WARN    dns configurator: could not resolve github.com (try 6 of 6): lookup github.com on 10.96.0.10:53: server misbehaving
2020-12-30T11:16:34.647-0600    INFO    dns over tls: falling back on plaintext DNS at address 10.96.0.10
2020-12-30T11:16:34.647-0600    INFO    dns configurator: using DNS address 10.96.0.10 internally
2020-12-30T11:16:34.647-0600    INFO    dns configurator: using DNS address 10.96.0.10 system wide
2020-12-30T11:16:34.647-0600    WARN    dns over tls: Unbound does not seem to be working after 6 tries
2020-12-30T11:16:34.647-0600    INFO    dns over tls: attempting restart in 10 seconds
```

2. Something weird is happening with `/etc/resolv.conf`. I'm seeing newly appending values continually. After a minute or so of the container failing to start, my conf is >5K lines long, which looks similar to below and repetitive. Eventually, this consumes all memory on the affected host.

```bash
/ # cat /etc/resolv.conf
search nzbget.svc.cluster.local svc.cluster.local cluster.local
nameserver 10.96.0.10
options ndots:5
search nzbget.svc.cluster.local svc.cluster.local cluster.local
nameserver 10.96.0.10
options ndots:5search nzbget.svc.cluster.local svc.cluster.local cluster.local
nameserver 127.0.0.1
options ndots:5
search nzbget.svc.cluster.local svc.cluster.local cluster.local
nameserver 127.0.0.1
options ndots:5search nzbget.svc.cluster.local svc.cluster.local cluster.local
nameserver 10.96.0.10
options ndots:5
search nzbget.svc.cluster.local svc.cluster.local cluster.local
nameserver 10.96.0.10
options ndots:5search nzbget.svc.cluster.local svc.cluster.local cluster.local
nameserver 10.96.0.10
options ndots:5
... TRUNCATED ...

/ # wc -l /etc/resolv.conf
5120 /etc/resolv.conf
```

**Share your logs...**

```log
Running version latest built on 2020-12-30T15:54:51Z (commit fb8a615)

📣  New Docker image qmcgaw/gluetun

🔧  Need help? https://github.com/qdm12/gluetun/issues/new
💻  Email? quentin.mcgaw@gmail.com
☕  Slack? Join from the Slack button on Github
💸  Help me? https://github.com/sponsors/qdm12
2020-12-30T11:16:26.503-0600    INFO    Unbound version: 1.10.1
2020-12-30T11:16:26.504-0600    INFO    IPtables version: v1.8.4
2020-12-30T11:16:26.507-0600    INFO    OpenVPN version: 2.4.9
2020-12-30T11:16:26.508-0600    INFO    Settings summary below:
OpenVPN settings:
|--User: [redacted]
|--Password: [redacted]
|--Verbosity level: 1
|--Run as root: no
|--Private Internet Access settings:
 |--Network protocol: tcp
 |--Regions: <removed>
 |--Encryption preset: strong
 |--Port forwarding: off
System settings:
|--Process user ID: 1000
|--Process group ID: 1000
|--Timezone: america/chicago
DNS over TLS disabled, using plaintext DNS 10.96.0.10
Firewall settings:
 |--VPN input ports:
 |--Input ports:
 |--Outbound subnets: 10.96.0.0/12
HTTP Proxy settings: disabled
ShadowSocks settings: disabled
HTTP Control server:
 |--Listening port: 8000
 |--Logging: true
Server updater settings: disabled
Public IP getter settings:
|--Period: 12h0m0s
|--IP file: /tmp/gluetun/ip
Version information: enabled

2020-12-30T11:16:26.508-0600    INFO    storage: creating /gluetun/servers.json with 6739 hardcoded servers
2020-12-30T11:16:26.527-0600    INFO    routing: default route found: interface eth0, gateway 10.42.0.0
2020-12-30T11:16:26.528-0600    INFO    routing: local subnet found: 10.32.0.0/12
2020-12-30T11:16:26.528-0600    INFO    routing: default route found: interface eth0, gateway 10.42.0.0
2020-12-30T11:16:26.528-0600    INFO    routing: adding route for 0.0.0.0/0
2020-12-30T11:16:26.528-0600    INFO    firewall: firewall disabled, only updating allowed subnets internal list
2020-12-30T11:16:26.528-0600    INFO    routing: default route found: interface eth0, gateway 10.42.0.0
2020-12-30T11:16:26.528-0600    INFO    routing: adding route for 10.96.0.0/12
2020-12-30T11:16:26.528-0600    INFO    openvpn configurator: checking for device /dev/net/tun
2020-12-30T11:16:26.528-0600    WARN    TUN device is not available: open /dev/net/tun: no such file or directory
2020-12-30T11:16:26.528-0600    INFO    openvpn configurator: creating /dev/net/tun
2020-12-30T11:16:26.528-0600    INFO    firewall: enabling...
2020-12-30T11:16:26.538-0600    INFO    firewall: enabled successfully
2020-12-30T11:16:26.538-0600    INFO    healthcheck: listening on 127.0.0.1:9999
2020-12-30T11:16:26.538-0600    INFO    dns over tls: using plaintext DNS at address 10.96.0.10
2020-12-30T11:16:26.538-0600    INFO    dns configurator: using DNS address 10.96.0.10 internally
2020-12-30T11:16:26.538-0600    INFO    dns configurator: using DNS address 10.96.0.10 system wide
2020-12-30T11:16:26.538-0600    INFO    http server: listening on 0.0.0.0:8000
2020-12-30T11:16:26.538-0600    INFO    firewall: setting VPN connection through firewall...
2020-12-30T11:16:26.538-0600    INFO    Launching standard output merger
2020-12-30T11:16:26.539-0600    INFO    openvpn configurator: starting openvpn
2020-12-30T11:16:26.541-0600    INFO    openvpn: OpenVPN 2.4.9 x86_64-alpine-linux-musl [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [MH/PKTINFO] [AEAD] built on Apr 20 2020
2020-12-30T11:16:26.541-0600    INFO    openvpn: library versions: OpenSSL 1.1.1i  8 Dec 2020, LZO 2.10
2020-12-30T11:16:26.541-0600    WARN    openvpn: --ping should normally be used with --ping-restart or --ping-exit
2020-12-30T11:16:26.544-0600    INFO    openvpn: CRL: loaded 1 CRLs from file [[INLINE]]
2020-12-30T11:16:26.545-0600    INFO    openvpn: TCP/UDP: Preserving recently used remote address: [AF_INET]154.3.250.24:501
2020-12-30T11:16:26.545-0600    INFO    openvpn: Attempting to establish TCP connection with [AF_INET]154.3.250.24:501 [nonblock]
2020-12-30T11:16:28.545-0600    INFO    openvpn: TCP connection established with [AF_INET]xxx.xxx.xxx.xxx:501
2020-12-30T11:16:28.545-0600    INFO    openvpn: TCP_CLIENT link local: (not bound)
2020-12-30T11:16:28.545-0600    INFO    openvpn: TCP_CLIENT link remote: [AF_INET]xxx.xxx.xxx.xxx:501
2020-12-30T11:16:28.791-0600    INFO    openvpn: [dallas415] Peer Connection Initiated with [AF_INET]xxx.xxx.xxx.xxx:501
2020-12-30T11:16:30.014-0600    INFO    openvpn: OpenVPN ROUTE6: OpenVPN needs a gateway parameter for a --route-ipv6 option and no default was specified by either --route-ipv6-gateway or --ifconfig-ipv6 options
2020-12-30T11:16:30.014-0600    INFO    openvpn: OpenVPN ROUTE: failed to parse/resolve route for host/network: 2000::/3
2020-12-30T11:16:30.014-0600    INFO    openvpn: TUN/TAP device tun0 opened
2020-12-30T11:16:30.014-0600    INFO    openvpn: /sbin/ip link set dev tun0 up mtu 1500
2020-12-30T11:16:30.019-0600    INFO    openvpn: /sbin/ip addr add dev tun0 10.21.19.2/24 broadcast 10.21.19.255
2020-12-30T11:16:30.022-0600    WARN    openvpn: OpenVPN was configured to add an IPv6 route over tun0. However, no IPv6 has been configured for this interface, therefore the route installation may fail or may not work as expected.
2020-12-30T11:16:30.022-0600    INFO    openvpn: UID set to nonrootuser
2020-12-30T11:16:30.022-0600    INFO    openvpn: Initialization Sequence Completed
2020-12-30T11:16:30.022-0600    INFO    dns configurator: downloading root hints from https://raw.githubusercontent.com/qdm12/files/master/named.root.updated
2020-12-30T11:16:30.360-0600    INFO    dns configurator: downloading root key from https://raw.githubusercontent.com/qdm12/files/master/root.key.updated
2020-12-30T11:16:30.431-0600    INFO    dns configurator: generating Unbound configuration
2020-12-30T11:16:30.431-0600    INFO    dns configurator: 0 hostnames blocked overall
2020-12-30T11:16:30.431-0600    INFO    dns configurator: 0 IP addresses blocked overall
2020-12-30T11:16:30.431-0600    INFO    dns configurator: starting unbound
2020-12-30T11:16:30.431-0600    INFO    dns configurator: using DNS address 127.0.0.1 internally
2020-12-30T11:16:30.431-0600    INFO    dns configurator: using DNS address 127.0.0.1 system wide
2020-12-30T11:16:30.488-0600    INFO    unbound: start of service (unbound 1.10.1).
2020-12-30T11:16:30.734-0600    WARN    dns configurator: could not resolve github.com (try 1 of 6): lookup github.com on 10.96.0.10:53: server misbehaving
2020-12-30T11:16:30.837-0600    WARN    dns configurator: could not resolve github.com (try 2 of 6): lookup github.com on 10.96.0.10:53: server misbehaving
2020-12-30T11:16:31.139-0600    WARN    dns configurator: could not resolve github.com (try 3 of 6): lookup github.com on 10.96.0.10:53: server misbehaving
2020-12-30T11:16:31.642-0600    WARN    dns configurator: could not resolve github.com (try 4 of 6): lookup github.com on 10.96.0.10:53: server misbehaving
2020-12-30T11:16:32.644-0600    WARN    dns configurator: could not resolve github.com (try 5 of 6): lookup github.com on 10.96.0.10:53: server misbehaving
2020-12-30T11:16:34.647-0600    WARN    dns configurator: could not resolve github.com (try 6 of 6): lookup github.com on 10.96.0.10:53: server misbehaving
2020-12-30T11:16:34.647-0600    INFO    dns over tls: falling back on plaintext DNS at address 10.96.0.10
2020-12-30T11:16:34.647-0600    INFO    dns configurator: using DNS address 10.96.0.10 internally
2020-12-30T11:16:34.647-0600    INFO    dns configurator: using DNS address 10.96.0.10 system wide
2020-12-30T11:16:34.647-0600    WARN    dns over tls: Unbound does not seem to be working after 6 tries
2020-12-30T11:16:34.647-0600    INFO    dns over tls: attempting restart in 10 seconds
2020-12-30T11:16:34.647-0600    INFO    VPN routing IP address: xxx.xxx.xxx.xxx
```

**Comment by qdm12** [comment]
Oooupsi in the word here. Will get at fixing it right now, thanks for reporting the issue! #329 duplicates this too.

**Comment by qdm12** [comment]
For 1. it should be fixed. I don't think it was necessarily the firewall, more of a [logic mess in my code](https://github.com/qdm12/gluetun/commit/c6eb5c1785937f7921f74f29a81453f72f2fa32d#diff-79794213df026c39b2fa937b850fe26e8891f30fadcc85d56a0c5bfdf5655da1) really...

For 2. this was due to my recent code changes (yesterday) where I reworked all the filesystem operations... and basically I was [opening the file as read write without truncating it](https://github.com/qdm12/gluetun/commit/f385c4203a33dabe845759eeca3b8927956f3db9#diff-fcb4cfb934ebc94181968551a8d94197cea8bbbfb23550886182d805bb4d2d3cR29)... hence the infinite file on retries; Plus I don't think DNS over TLS was even active as the Docker DNS was still active 😕  Anyway that's **also fixed** now 👍 

Let me know if anything doesn't work in your setup, I'll close the issue for now but I'm happy to re-open it!
